File: backend/src/test/java/com/skewwhiffy/auraltester/test/util/TestData.java
package com.skewwhiffy.auraltester.test.util;

import com.skewwhiffy.auraltester.notation.factory.*;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import com.skewwhiffy.auraltester.notation.model.interval.DirectedInterval;
import com.skewwhiffy.auraltester.notation.model.interval.Interval;
import com.skewwhiffy.auraltester.notation.model.key.Key;
import com.skewwhiffy.auraltester.notation.model.note.*;
import com.skewwhiffy.auraltester.notation.model.scale.Scale;
import com.skewwhiffy.auraltester.notation.model.scale.ScaleDirection;
import com.skewwhiffy.auraltester.notation.model.scale.ScaleType;

import java.text.MessageFormat;
import java.util.*;
import java.util.stream.IntStream;

public class TestData {
    public static RandomTestData random() {
        return new RandomTestData();
    }

    public static NoteFactories noteFactories() {
        return new NoteFactories();
    }

    public static class RandomTestData {
        private final Random random = new Random();

        public String string() {
            return MessageFormat.format("{0}", UUID.randomUUID());
        }

        private String oneOf(String source) {
            List<String> charList = source
                    .chars()
                    .mapToObj(it -> (char) it)
                    .map(Object::toString)
                    .toList();
            return oneOf(charList);
        }

        public <T> T oneOf(T... source) {
            return oneOf(Arrays.stream(source).toList());
        }

        public <T> T oneOf(List<T> source) {
            return source.get(random.nextInt(source.size()));
        }

        public AbsoluteNote absoluteNote() {
            return new AbsoluteNote(note(), octave(), Optional.empty());
        }

        private Accidental accidental() {
            return new Accidental(random.nextInt(3) - 1);
        }

        public Clef clef() {
            return new Clef(string(), absoluteNote(), absoluteNote());
        }

        public DirectedInterval directedInterval() {
            return random.nextBoolean() ? interval().up() : interval().down();
        }

        public Interval interval() {
            return new Interval(
                    oneOf(IntStream.range(1, 9).boxed().toList()),
                    random.nextInt(3) - 1
            );
        }

        public IntervalNotes intervalNotes() {
            return new IntervalNotes(absoluteNote(), interval());
        }

        public Key key() {
            return random.nextBoolean() ? Key.major(note()) : Key.minor(note());
        }

        public Note note() {
            return new Note(oneOf("ABCDEFG"), accidental());
        }

        private Octave octave() {
            return new Octave(random.nextInt(8) - 4);
        }

        public Scale scale() {
            return new Scale(absoluteNote(), scaleType(), scaleDirection());
        }

        public ScaleDirection scaleDirection() {
            return oneOf(Arrays.stream(ScaleDirection.values()).toList());
        }

        public ScaleType scaleType() {
            return new ScaleType(
                    string(),
                    string(),
                    IntStream.range(1, 6).mapToObj(it -> directedInterval()).toList(),
                    IntStream.range(1, 6).mapToObj(it -> directedInterval()).toList()
            );
        }
    }

    public static class NoteFactories {
        public NoteFactory note() {
            return new NoteFactory();
        }

        public ClefFactory clef() {
            return new ClefFactory(note());
        }

        public IntervalFactory interval() {
            return new IntervalFactory();
        }

        public KeyFactory key() {
            return new KeyFactory(note());
        }

        public InternalNotationFactory internalNotation() {
            return new InternalNotationFactory(
                    clef(),
                    note(),
                    interval(),
                    key()
            );
        }

        public ScaleTypeFactory scaleType() {
            return new ScaleTypeFactory(interval());
        }
    }

}

File: backend/src/test/kotlin/com/skewwhiffy/auraltester/test/util/TestData.kt
package com.skewwhiffy.auraltester.test.util

import com.skewwhiffy.auraltester.controller.ClefType
import com.skewwhiffy.auraltester.helper.oneOf
import com.skewwhiffy.auraltester.notation.factory.ClefFactory
import com.skewwhiffy.auraltester.notation.factory.InternalNotationFactory
import com.skewwhiffy.auraltester.notation.factory.IntervalFactory
import com.skewwhiffy.auraltester.notation.factory.KeyFactory
import com.skewwhiffy.auraltester.notation.factory.NoteFactory
import com.skewwhiffy.auraltester.notation.factory.ScaleTypeFactory
import com.skewwhiffy.auraltester.notation.model.abc.ignoreAccidental
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.notation.model.interval.Interval
import com.skewwhiffy.auraltester.notation.model.key.Key
import com.skewwhiffy.auraltester.notation.model.note.*
import com.skewwhiffy.auraltester.notation.model.scale.Scale
import com.skewwhiffy.auraltester.notation.model.scale.ScaleDirection
import com.skewwhiffy.auraltester.notation.model.scale.ScaleType
import java.util.*

class TestData {
    companion object {
        val random: RandomTestData by lazy(::RandomTestData)
        val noteFactories by lazy(::NoteFactories)

        class NoteFactories {
            val note by lazy(::NoteFactory)
            val clef by lazy { ClefFactory(note) }
            private val interval by lazy(::IntervalFactory)
            val key by lazy { KeyFactory(note) }
            val internalNotation by lazy { InternalNotationFactory(clef, note, interval, key) }
            val scaleType by lazy { ScaleTypeFactory(interval) }
        }
    }

    class RandomTestData {
        private val random: Random by lazy(::Random)

        val string
            get() = UUID.randomUUID().toString()

        val absoluteNote
            get() = AbsoluteNote(note, octave, null)

        private val accidental
            get() = Accidental(random.nextInt(3) - 1)

        val clef
            get() = Clef(
                clefType,
                string.uppercase(),
                absoluteNote.ignoreAccidental,
                absoluteNote.ignoreAccidental
            )

        val clefType
            get() = oneOf(ClefType.entries)

        val directedInterval
            get() = if (random.nextBoolean()) interval.up else interval.down

        val interval
            get() = Interval(oneOf((1..8)), random.nextInt(3) - 1)

        val intervalNotes
            get() = IntervalNotes(absoluteNote, interval)

        val key
            get() = if (random.nextBoolean()) Key.major(note) else Key.minor(note)

        @Suppress("SpellCheckingInspection")
        val note
            get() = Note(oneOf("ABCDEFG"), accidental)

        private val octave
            get() = Octave(random.nextInt(8) - 4)

        val scale
            get() = Scale(absoluteNote, scaleType, scaleDirection)

        val scaleDirection
            get() = oneOf(ScaleDirection.entries.toList())

        val scaleType
            get() = ScaleType(
                string,
                string,
                (1..5).map { directedInterval },
                (1..5).map { directedInterval }
            )
    }
}
